### PR TITLE
Rework the API for move to the `bridgetree` crate.

### PR DESCRIPTION
**Pull request by nuttycom**
This builds atop #43. It clarifies the naming of a number of operations, and moves the `Frontier` and `Tree` traits to a `testing` module while moving the `BridgeTree` type to the top level.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,1556 +1,1510 @@
...
-        (self.0..other.0).into_iter().map(Level)
-    }
+use std::collections::{BTreeMap, BTreeSet};
+use std::convert::TryFrom;
+use std::fmt::Debug;
+use std::mem::size_of;
+use std::ops::Range;
+
+use crate::{
+    hashing::Hashable,
```
Comment (ebfull) [inline src/lib.rs:47]: We should `pub use` this trait at the `lib.rs` level, and not expose the `hashing` module publicly.

File: src/lib.rs [hunk 2]
```diff
@@ -1,1556 +1,1510 @@
...
+    PositionNotMarked(Position),
+    BridgeFusionError(ContinuityError),
+    BridgeAddressInvalid(Address),
 }
 
-impl From<Level> for u8 {
-    fn from(level: Level) -> u8 {
-        level.0
-    }
+/// A `[NonEmptyFrontier]` is a reduced representation of a Merkle tree,
```
Comment (ebfull) [inline src/lib.rs:86]:
```suggestion
/// A [`NonEmptyFrontier`] is a reduced representation of a Merkle tree,
```

File: src/lib.rs [hunk 3]
```diff
@@ -1,1556 +1,1510 @@
...
     }
 }
 
-impl From<Position> for Address {
-    fn from(p: Position) -> Self {
-        Address {
-            level: 0.into(),
-            index: p.into(),
+impl<'a, H: Hashable + Ord + Clone + 'a> MerkleBridge<H> {
+    /// Constructs a new bridge to follow this one. If mark_current_leaf is true, the successor
```
Comment (ebfull) [inline src/lib.rs:425]:
```suggestion
    /// Constructs a new bridge to follow this one. If `mark_current_leaf` is true, the successor
```

File: src/lib.rs [hunk 4]
```diff
@@ -1,1556 +1,1510 @@
...
+}
 
-    /// Remove state from the tree that no longer needs to be maintained
-    /// because it is associated with checkpoints or witnesses that
-    /// have been removed from the tree at positions deeper than those
-    /// reachable by calls to `rewind`. It is always safe to implement
-    /// this as a no-op operation
-    fn garbage_collect(&mut self);
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct BridgeTree<H, const DEPTH: u8> {
```
Comment (ebfull) [inline src/lib.rs:661]: This structure isn't documented -- probably should be at least a little considering it's the whole point of the crate. :P

File: src/lib.rs [hunk 5]
```diff
@@ -1,1556 +1,1510 @@
...
-#[derive(Copy, Clone, Debug, PartialEq, Eq)]
-pub(crate) enum Source {
-    /// The sibling to the address can be derived from the incremental frontier
-    /// at the contained ommer index
-    Past(usize),
-    /// The sibling to the address must be obtained from values discovered by
-    /// the addition of more nodes to the tree
-    Future,
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct MerkleBridge<H> {
```
Comment (ebfull) [inline src/lib.rs:314]: This structure is also not documented.

**Review by ebfull — APPROVED** [review]
ACK modulo (mostly) nits

File: src/lib.rs (hunk 1 above)
Comment (nuttycom) [reply src/lib.rs:47]: Hah, yeah, I noticed this when using the crate in librustzcash.

**Review by ebfull — APPROVED** [review]

**Review by ebfull — COMMENTED** [review]
post-hoc ACK

**Pull request merged**